src/core/include/units/magnitude.h

```diff
@@ -0,0 +1,360 @@
+
+/**
+ * @brief  A representation for positive real numbers which optimizes taking products and rational powers.
+ *
+ * Magnitudes can be treated as values.  Each type encodes exactly one value.  Users can multiply, divide, raise to
+ * rational powers, and compare for equality.
+ */
+template<BasePower auto... BPs>
+  requires ((detail::is_valid_base_power(BPs) && ... && detail::strictly_increasing(BPs.get_base()...)))
+struct magnitude {};
+
+// Implementation for Magnitude concept (below).
+namespace detail {
+template<typename T>
+static constexpr bool is_magnitude = false;
+template<BasePower auto... BPs>
+static constexpr bool is_magnitude<magnitude<BPs...>> = true;
+} // namespace detail
+
+/**
+ * @brief  Concept to detect whether T is a valid Magnitude.
+ */
+template<typename T>
+concept Magnitude = detail::is_magnitude<T>;
+
+/**
+ * @brief  Convert any positive integer to a Magnitude.
+ *
+ * This will be the main way end users create Magnitudes.  They should rarely (if ever) create a magnitude<...> by
+ * manually adding base powers.
+ */
+template<ratio R>
+  requires (R.num > 0)
+constexpr Magnitude auto as_magnitude();
+
+/**
+ * @brief  A base to represent pi.
+ */
+struct pi_base {
+  static constexpr long double value = std::numbers::pi_v<long double>;
+};
+
+/**
+ * @brief  A simple way to create a Magnitude representing a rational power of pi.
+ */
+template<ratio Power>
+constexpr auto pi_to_the() { return magnitude<base_power<pi_base>{Power}>{}; }
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Magnitude equality implementation.
+
+template<BasePower auto... LeftBPs, BasePower auto... RightBPs>
+constexpr bool operator==(magnitude<LeftBPs...>, magnitude<RightBPs...>) {
+  if constexpr (sizeof...(LeftBPs) == sizeof...(RightBPs)) { return ((LeftBPs == RightBPs) && ...); }
+  else { return false; }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Magnitude rational powers implementation.
+
+template<ratio E, BasePower auto... BPs>
+constexpr auto pow(magnitude<BPs...>) {
+  if constexpr (E == 0) { return magnitude<>{}; }
+  else { return magnitude<pow<E>(BPs)...>{}; }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Magnitude product implementation.
+
+// Base cases, for when either (or both) inputs are the identity.
+constexpr auto operator*(magnitude<>, magnitude<>) { return magnitude<>{}; }
+constexpr auto operator*(magnitude<>, Magnitude auto m) { return m; }
+constexpr auto operator*(Magnitude auto m, magnitude<>) { return m; }
+
+// Recursive case for the product of any two non-identity Magnitudes.
+template<BasePower auto H1, BasePower auto... T1, BasePower auto H2, BasePower auto... T2>
+constexpr auto operator*(magnitude<H1, T1...>, magnitude<H2, T2...>) {
```

Comment:
Operators could be implemented as hidden friends.

Reply:
Good point!  I typically prefer hidden friends, but for some reason it hadn't occurred to me here.  I started working on this, but it seemed like I'd need multiple specializations of `magnitude` for different patterns of parameter packs.  If that's true, I'm not sure it's worth the added complexity.  But maybe there's a strategy for this which I didn't see.

Reply:
Let's merge it as it is and then possibly try to refactor.